docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,196 @@
+federated servers.
+
+## Motivation
+
+* Extensibility: We want to allow community members to write their own API
+  servers to expose APIs they want. Cluster admins should be able to use these
+  servers without having to require any change in the core kubernetes
+  repository.
+* Unblock new APIs from core kubernetes team review: A lot of new API proposals
+  are currently blocked on review from the core kubernetes team. By allowing
+  developers to expose their APIs as a separate server and enabling the cluster
+  admin to use it without any change to the core kubernetes reporsitory, we
+  unblock these APIs.
+* Place for staging experimental APIs: New APIs can remain in seperate
+  federated servers until they become stable, at which point, they can be moved
+  to the core kubernetes master, if appropriate.
+
+## Goal
+
+* Developers should be able to write their own API server and cluster admins
+  should be able to add them to their cluster, exposing new APIs at runtime. All
+  of this should not require any change to the core kubernetes API server.
+* These new APIs should be seamless extension of the core kubernetes APIs (ex:
+  they should be operated upon via kubectl).
+
+## Non Goals:
+
+The following are related but are not the goals of this specific proposal:
+* Make it easy to write a kubernetes API server.
+
+## High Level Architecture
+
+There will be a reverse proxy in front of all API servers. Clients will talk to
+the proxy to discover the API groups and resources exposed by all of them.
+Clients can use the proxy to talk to federated servers, without having to know
+their individual URLs. They can talk to each server directly as well (without
+going through the proxy), if desired, by getting server URLs from discovery API.
+
+### Constraints
+
+* Unique API groups across servers: Each API server (and groups of servers, in HA)
+  should expose unique API groups.
+* Follow API conventions: APIs exposed by every API server should adhere to [kubernetes API
+  conventions](../devel/api-conventions.md).
+* Support discovery API: Each API server should support the kubernetes discovery API
+  (list the suported groupVersions at `/apis` and list the supported resources
+  at `/apis/<groupVersion>/`)
+* No bootstrap problem: The core kubernetes server should not depend on any
+  other federated server to come up. Other servers can only depend on the core
+  kubernetes server.
+
+## Implementation Details
+
+### Summarizing discovery information
+
+We can have a very simple Go program to summarize discovery information from all
+servers. Each federated API server will register its baseURL and swagger spec path
+with it. It will summarize the list of all group versions exposed by all registered API
+servers, with their individual URLs at `/apis`. Clients can then use the server
+URL to talk to that server directly.
+
+### Reverse proxy
+
+The same Go program can be extended to act as reverse proxy for all federated
```

Comment:
Instead of extending it, it'd be easier to just use an existing proxy. Since given that go program, there's no longer any special thing that the proxy would need.